llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -11807,10 +11824,10 @@ class BoUpSLP::ShuffleInstructionBuilder final : public BaseShuffleAnalysis {
   Value *castToScalarTyElem(Value *V,
                             std::optional<bool> IsSigned = std::nullopt) {
     auto *VecTy = cast<VectorType>(V->getType());
-    if (VecTy->getElementType() == ScalarTy)
+    if (VecTy->getElementType() == ScalarTy->getScalarType())
```

Comment:
What if the number of elements in ScalarTy and in VecTy is different, is this possible?

Reply:
Then you need to handle it here or add the assertion that the number of elements expected to be the same only

Reply:
Maybe my previous description cause misleading.
The `V` here is a vectorized result. The number of elements for `V` cannot be equal to the number of elements for `ScalarTy`.
For example,
```
  %0 = load <4 x i32>, ptr %a, align 4
  %1 = load <4 x i32>, ptr %arrayidx3, align 4
  %2 = load <4 x i32>, ptr %arrayidx7, align 4
  %3 = load <4 x i32>, ptr %arrayidx11, align 4
```
The `ScalarTy` here is `<4 x i32>` but `VecTy` is `<16 x i32>`. (from `FinalShuffle -> ShuffleBuilder.addOrdered -> add -> castToScalarTyElem`)
It also happend when REVEC is disabled. See `llvm/test/Transforms/SLPVectorizer/reschedule.ll`, `ScalarTy` is `double` but `VecTy` is `<2 x double>`.

Reply:
Can you add the assertion, that ScalarTy is either the scalar or has the same number of elements as VecTy?